neumai/neumai/SinkConnectors/MarqoSink.py

```diff
@@ -0,0 +1,187 @@
+        '''
+        if recreate_index:
+            if index_name in [i.index_name for i in marqo_client.get_indexes()['results']]:
+                marqo_client.delete_index(index_name)
+
+        if not index_name:
+            raise Exception('Index name must not be none')
+
+        if similarity not in ['cosinesimil', 'l1', 'l2', 'linf']:
+            raise Exception('Similarity function must be one of l1, l2, linf and cosinesimil')
+
+        if index_name in [i.index_name for i in marqo_client.get_indexes()['results']] and recreate_index==False:
+            raise Exception(f'Index {index_name} already exists, please use another name')
+
+        # Not using any model, providing own vectors
+        # https://docs.marqo.ai/1.4.0/API-Reference/Indexes/create_index/#no-model
+        marqo_client.create_index(
+        index_name=index_name,
+        settings_dict={
+                    'index_defaults': {
+                        'model': 'no_model',
+                        'model_properties': {
+                            'dimensions': embedding_dim
+                        },
+                        'ann_parameters':{
+                            'space_type': similarity
+                        }
+                    }
+                }
+        )
+
+    def store(self, vectors_to_store:List[NeumVector]) -> int:
+        url = self.url
+        api_key = self.api_key
+        index_name = self.index_name
+
+        marqo_client = marqo.Client(
+            url=url, 
+            api_key=api_key,
+        )
+        self._create_index(index_name=index_name,
+                          marqo_client=marqo_client,
+                          similarity="cosinesimil",
+                          embedding_dim=len(vectors_to_store[0].vector))
+        mod_vecs = []
+        for vec in vectors_to_store:
+            _id = vec.id
+            _vec = vec.vector
+            dic = {
+                    '_id': _id,
+                    'neum': {
+                        'vector': _vec
+                    }
+                }
+            # Using the two lines below, we can unroll the elements in metadata
+            # and add them separately
+            for k,v in vec.metadata.items():
+                dic[k] = v
+            # dic['metadata'] = vec.metadata
+            mod_vecs.append(
+                dic
+            )
+        operation_info = marqo_client.index(index_name).add_documents(
+            documents=mod_vecs,
+            mappings={
+                'neum': 
+                    {
+                    'type': 'custom_vector'
+                    }
+            },
+            tensor_fields=['neum'],
+            auto_refresh=True
+            )
+        
+        if(operation_info['errors'] == False):
+            return  len(operation_info['items'])
+        raise MarqoInsertionException("Marqo storing failed. Try again later.")
+    
```

Comment:
```suggestion
    
 def delete_vectors_with_file_id(self, file_id: str) -> bool:
        raise Exception("Marqo doesn't have support to delete vectors by metadata")
    
```

Reply:
In the recent 0.0.35 version we added delete capabilities. Please add support in the PR. If Marqo doesn't have support then just mark as not supported. If it does then implement. 

Delete is based on a metadata object for file_id and is designed to remove all the vectors for a given file.

Reply:
Well, marqo does support deletion, but its based on document id, can you elaborate what you mean by a `file` here, is it one object with many fields which are vectorized and you want to delete those vectors for that object?

Reply:
Correct, the delete implementation we have is currently based on deleting all the vectors associated to a given file.

Reply:
If deletion can only be done based on the vector id, then we might need to think about how to implement this. 

For other DBs we are simply basing the deletion on a filter.

Reply:
@ddematheu I got it, I think marqo has what you need, you can give a document id and delete all vectors associated with it, for exampe - if your doc had three fields, namely `title`, `text` and `author`, and you had vectorized `title` and `text`, when you delete the doc, it would delete all vectors(both title and text).